test/e2e/proxy/applications_test.go

```diff
@@ -0,0 +1,361 @@
+package proxy
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	. "github.com/codeready-toolchain/toolchain-e2e/testsupport"
+	appstudiov1 "github.com/codeready-toolchain/toolchain-e2e/testsupport/appstudio/api/v1alpha1"
+	. "github.com/codeready-toolchain/toolchain-e2e/testsupport/proxy"
+	"github.com/codeready-toolchain/toolchain-e2e/testsupport/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// full flow from usersignup with approval down to namespaces creation and cleanup
+//
+// !!! Additional context !!!
+// The test uses a dummy HAS API type called Application. The reason for creating the dummy
+// Application CRD instead of using the real one is that we want to reduce the adherence
+// to 3rd party operators (we don't want to have to install the operator for our e2e tests).
+// To work around this limitation, we created a dummy HAS API type that has the same name
+// and the same group as the actual one. The CRD is created as part of the test setup
+// and since the CRD name & group name matches, then RBAC allow us to execute create/read
+// operations on that resource using the user permissions.
+func TestProxyFlow(t *testing.T) {
```

Comment:
related to the other comment, I don't see how we could get closer to sending many requests to proxy in a parallel way. There are many sub-tests that will be executed in sequential order.
I hoped that we could have something closer to what we do in migration tests. see 
https://github.com/codeready-toolchain/toolchain-e2e/blob/713b9018786877b8782f3aecd0c3744616d50382/test/migration/setup_runner.go#L38-L59
https://github.com/codeready-toolchain/toolchain-e2e/blob/713b9018786877b8782f3aecd0c3744616d50382/test/migration/verify/verify_migration_test.go#L68-L106

A very very quick example of what it could look like:
```go
...
usersToBeTested := createUsersForParallelTests("user1"..."user10")
usersToBeTested["user1"].ShareSpaceWith(t, hostAwait, users["user2"])
usersToBeTested["user4"].ShareSpaceWith(t, hostAwait, users["user2"])
usersToBeTested["user10"].ShareSpaceWith(t, hostAwait, users["user5"])

...

for _, user := range usersToBeTested {
	// provision users
	user.provision()
	user.shareWithOtherUsers()
	user.verify().assessToWorkspaces() // space lister verification

	for _, action := actionsToBeTested { // CRUD, watch, websockets operations, etc...
		go func() {
			user.verify().action(action).inAllWorkspaces()
		}()
	}
}
...
```
the actions to be tested are the whole stories like
action 1:
1. create CM
2. list CMs
3. delete CM

action2
1. watch secrets
2. create secret
3. modify it
4. delete it

action3
1. create pod
5. see the logs
6. delete pod

etc...

All these things could be easily hidden behind a nice fluent API and set of functions. You can technically reuse most of the functions that are already there - eg the one for running websockets.

Is this still your plan and this PR is only a preparation for it?

Reply:
Thank you for this comment. I am working on running tests in parallel in similar way, as you pointed out and this PR is more about preparation for next PR(s) with parallelization.